### PR TITLE
feat: add jaeger rest endpoints for grafana tracing support

### PR DESCRIPTION
**Pull request by etolbakov**
### Description

Add jaeger rest endpoints for grafana tracing support

TODO:

- [x] Remove `go_parse_duration` dependency
- [x] Move functions in `util.rs` into `models.rs` (can we get rid of some of them?)
- [x] Check if the swagger UI is working correctly with the new endpoints

### Related issues:
https://github.com/quickwit-oss/quickwit-datasource/issues/5

### CODE REVIEW

File: quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+use crate::{with_arg, BodyFormat};
+
+/// Setup Jaeger API handlers
+///
+/// This is where all Jaeger handlers
+/// should be registered.
+pub(crate) fn jaeger_api_handlers(
+    jaeger_service_opt: Option<JaegerService>,
+) -> impl Filter<Extract = (impl warp::Reply,), Error = Rejection> + Clone {
+    jaeger_services_handler(jaeger_service_opt.clone())
```
Comment (etolbakov) [inline quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs:46]: should I pass `Option<JaegerService>`  further down or it's better to unwrap Option here and respond with something like `internal_jaeger_error()` for the `None` case?

File: quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+#[utoipa::path(
+    get,
+    tag = "Jaeger Services",
+    path = "/services",
+    responses(
+        (status = 200, description = "Successfully fetched services information.", body = TODO )
+    )
+)]
+pub(crate) fn jaeger_services_filter() -> impl Filter<Extract = (), Error = Rejection> + Clone {
+    warp::path!("jaeger" / "services").and(warp::get())
```
Comment (etolbakov) [inline quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs:101]: should this path be changed to `otel-traces-v0_6/services`?

Comment (etolbakov) [reply quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs:101]: already made an adjustment, will push it as a part of the next chunk

File: quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+#[utoipa::path(
+    get,
+    tag = "Jaeger Services",
+    path = "/services",
+    responses(
+        (status = 200, description = "Successfully fetched services information.", body = TODO )
+    )
+)]
+pub(crate) fn jaeger_services_filter() -> impl Filter<Extract = (), Error = Rejection> + Clone {
+    warp::path!("otel-traces-v0_6" / "services").and(warp::get())
```
Comment (fmassot) [inline quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs:102]: 2 issues:
- "services" should be prefix by "api" 
- we need to add a jaeger prefix to isolate those endpoints better.
 
So URL should be  "otel-traces-v0_6/jaeger/api/services"

I don't if it's possible but we could probably add a global prefix "otel-traces-v0_6/jaeger" in the parent handler

Comment (fmassot) [reply quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs:102]:
```rust
    let jaeger_api_root_url = warp::path!("otel-traces-v0_6" / "jaeger" / ..);
    jaeger_api_root_url.and(
        jaeger_services_handler(jaeger_service_opt.clone())
        .or(jaeger_service_operations_handler(
            jaeger_service_opt.clone(),
        ))
        .or(jaeger_traces_search_handler(jaeger_service_opt.clone()))
        .or(jaeger_traces_handler(jaeger_service_opt.clone()))
    )

```

Comment (fmassot) [reply quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs:102]: we should even go for `warp::path!(OTEL_TRACES_INDEX_ID / "jaeger" / "api" / ..);`

Comment (etolbakov) [reply quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs:102]: naive usage of the `const` gave me the following error
```
error[E0747]: constant provided when a type was expected
   --> quickwit-serve/src/jaeger_api/rest_handler.rs:104:17
    |
104 |     warp::path!(OTEL_TRACES_INDEX_ID / "jaeger" / "api" / "services").and(warp::get())
    |                 ^^^^^^^^^^^^^^^^^^^^

error[E0271]: type mismatch resolving `<And<And<And<And<And<And<impl Filter + FilterBase<Extract = (), Error = Infallible> + Copy, impl Filter + FilterBase<Extract = (...,), Error = ...> + Copy>, ...>, ...>, ...>, ...>, ...> as FilterBase>::Extract == ()`
   --> quickwit-serve/src/jaeger_api/rest_handler.rs:103:43
    |
103 | pub(crate) fn jaeger_services_filter() -> impl Filter<Extract = (), Error = Rejection> + Clone {
    |                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `(_,)`, found `()`
    |
    = note:  expected tuple `(_,)`
            found unit type `()`

Some errors have detailed explanations: E0271, E0747.
For more information about an error, try `rustc --explain E0271`.
error: could not compile `quickwit-serve` (lib) due to 2 previous errors
```
I tried a couple of things (specifying static lifetime, wrapping in the `Box`) and nothing worked.

Comment (etolbakov) [reply quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs:102]: agreed to proceed without `OTEL_TRACES_INDEX_ID`

File: quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+        .then(jaeger_services)
+        .map(|result| make_jaeger_api_response(result, BodyFormat::default()))
+}
+
+#[utoipa::path(
+    get,
+    tag = "Jaeger Operations",
+    path = "/otel-traces-v0_6/jaeger/api/services/{service}/operations",
+    responses(
+        (status = 200, description = "Successfully fetched operations data for the specified service.", body = JaegerResponseBody )
```
Comment (fmassot) [inline quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs:78]:
```suggestion
        (status = 200, description = "Successfully fetched operations names  the given service.", body = JaegerResponseBody )
```

File: quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+    get,
+    tag = "Jaeger Traces",
+    path = "/otel-traces-v0_6/jaeger/api/traces?service={service}&start={start_in_ns}&end={end_in_ns}&lookback=custom",
+    responses(
+        (status = 200, description = "Successfully fetched traces information.", body = JaegerResponseBody )
+    ),
+    params(
+        TracesSearchQueryParams,
+        ("service" = Option<String>, Query, description = "The service name."),
+        ("start" = Option<i64>, Query, description = "The start time."),
```
Comment (fmassot) [inline quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs:101]:
```suggestion
        ("start" = Option<i64>, Query, description = "The start time in nanosecond."),
```

File: quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+        ("end" = Option<i64>, Query, description = "The end time."),
```
Comment (fmassot) [inline quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs:102]:
```suggestion
        ("end" = Option<i64>, Query, description = "The end time in nanosecond."),
```

File: quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+        .then(jaeger_traces_search)
+        .map(|result| make_jaeger_api_response(result, BodyFormat::default()))
+}
+
+#[utoipa::path(
+    get,
+    tag = "Jaeger Traces",
+    path = "/otel-traces-v0_6/jaeger/api/traces/{id}/",
+    responses(
+        (status = 200, description = "Successfully fetched traces information for the provided id.", body = JaegerResponseBody )
```
Comment (fmassot) [inline quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs:121]:
```suggestion
        (status = 200, description = "Successfully fetched traces spans for the provided trace ID", body = JaegerResponseBody )
```

File: quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+    })
+}
+
+async fn jaeger_traces_search(
+    search_params: TracesSearchQueryParams,
+    jaeger_service: JaegerService,
+) -> Result<JaegerSearchBody, JaegerError> {
+    let query = TraceQueryParameters {
+        service_name: search_params.service.unwrap_or_default(),
+        operation_name: "stage_splits".to_string(),
```
Comment (fmassot) [inline quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs:178]: By looking at jaeger http requests, you can either specify operation "All" or the name of one operation. In the last case here is the HTTP request:

"http://localhost:16686/api/traces?end=1701555639361000&limit=20&lookback=3h&maxDuration&minDuration&operation=find_trace_ids&service=quickwit&start=1701544839361000"

**Comment by etolbakov** [comment]
Checked the swagger UI, the image below
image001

File: quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs [hunk 9]
```diff
@@ -114,8 +115,13 @@ pub fn jaeger_service_operations_handler(
     params(
         TracesSearchQueryParams,
         ("service" = Option<String>, Query, description = "The service name."),
+        ("operation" = Option<String>, Query, description = "The operation name."),
         ("start" = Option<i64>, Query, description = "The start time in nanoseconds."),
         ("end" = Option<i64>, Query, description = "The end time in nanoseconds."),
+        ("tags" = Option<String>, Query, description = "Sets tags with values in the logfmt format, such as error=true status=200."),
```
Comment (fmassot) [inline quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs:122]: Great!

File: quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs [hunk 10]
```diff
@@ -397,6 +407,7 @@ mod tests {
             .path(
                 "/otel-traces-v0_6/jaeger/api/traces?service=quickwit&\
                  operation=delete_splits_marked_for_deletion&minDuration=500us&maxDuration=1.2s&\
+                 tags=%7B%22tag.first%22%3A%22common%22%2C%22tag.second%22%3A%22true%22%7D&\
```
Comment (fmassot) [inline quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs:411]: we should check that the parameters are correctly built on the search request

File: quickwit/quickwit-serve/src/jaeger_api/model.rs [hunk 11]
```diff
@@ -57,6 +57,7 @@ pub struct TracesSearchQueryParams {
     pub operation: Option<String>,
     pub start: Option<i64>,
     pub end: Option<i64>,
+    pub tags: Option<String>,
```
Comment (fmassot) [inline quickwit/quickwit-serve/src/jaeger_api/model.rs:60]: not sure if this is the best way, maybe a `HashMap<String, serde_json:Value>` could help

File: quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs (hunk 10 above)
Comment (etolbakov) [reply quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs:411]: noted

Comment (etolbakov) [reply quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs:411]: @fmassot is it something you are after?
```
.withf(|req|
                {
                    assert!(req.query_ast.contains("{\"type\":\"term\",\"field\":\"resource_attributes.tag.first\",\"value\":\"common\"}"));
                    assert!(req.query_ast.contains("{\"type\":\"term\",\"field\":\"resource_attributes.tag.second\",\"value\":\"true\"}"));
                    assert_eq!(req.index_id_patterns, vec![OTEL_TRACES_INDEX_ID.to_string()]);
                    true
                }
            )
```

File: quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs [hunk 12]
```diff
@@ -0,0 +1,453 @@
...
+) -> Result<JaegerResponseBody<Vec<JaegerTrace>>, JaegerError> {
+    let duration_min = search_params
+        .min_duration
+        .map(parse_duration_with_units)
+        .transpose()?;
+    let duration_max = search_params
+        .max_duration
+        .map(parse_duration_with_units)
+        .transpose()?;
+    let tags = search_params
```
Comment (etolbakov) [inline quickwit/quickwit-serve/src/jaeger_api/rest_handler.rs:204]:
```suggestion
    let tags_opt = search_params
        .tags
        .clone()
        .map(|s| {
            serde_json::from_str::<HashMap<String, String>>(&s).map_err(|error| {
                error!(error = ?error, "failed to convert tags `{:?}`",search_params.tags.clone());
                JaegerError {
                    status: StatusCode::INTERNAL_SERVER_ERROR,
                    message: "failed to to convert tags".to_string(),
                }
            })
        })
        .transpose()?;
    let tags = tags_opt.unwrap_or(Default::default());
```

**Review by fmassot — APPROVED** [review]

**Comment by etolbakov** [comment]
😅🎉

**Pull request merged**